fenix/app/src/main/java/org/mozilla/fenix/tabstray/browser/compose/ReorderableTray.kt

```diff
@@ -0,0 +1,531 @@
+ * @param onExitLongPress Optional callback to be invoked when the item is dragged after long press
+ */
+@Composable
+fun rememberGridReorderState(
+    gridState: LazyGridState,
+    onMove: (LazyGridItemInfo, LazyGridItemInfo) -> Unit,
+    ignoredItems: List<Any>,
+    onLongPress: (LazyGridItemInfo) -> Unit = {},
+    onExitLongPress: () -> Unit = {},
+): GridReorderState {
+    val scope = rememberCoroutineScope()
+    val touchSlop = LocalViewConfiguration.current.touchSlop
+    val state = remember(gridState) {
+        GridReorderState(
+            gridState = gridState,
+            onMove = onMove,
+            scope = scope,
+            touchSlop = touchSlop,
+            ignoredItems = ignoredItems,
+            onLongPress = onLongPress,
+            onExitLongPress = onExitLongPress,
+        )
+    }
+    return state
+}
+
+/**
+ * Class containing details about the current state of dragging in grid
+ *
+ * @param gridState State of the grid
+ * @param scope [CoroutineScope] used for scrolling to the target item
+ * @param onMove Callback to be invoked when switching between two items
+ * @param onLongPress Optional callback to be invoked when long pressing an item
+ * @param onExitLongPress Optional callback to be invoked when the item is dragged after long press
+ * @param touchSlop Distance in pixels the user can wander until
+ * @param ignoredItems List of keys for non-draggable items
+ */
+class GridReorderState internal constructor(
+    private val gridState: LazyGridState,
+    private val scope: CoroutineScope,
+    private val onMove: (LazyGridItemInfo, LazyGridItemInfo) -> Unit,
+    private val onLongPress: (LazyGridItemInfo) -> Unit = {},
+    private val onExitLongPress: () -> Unit = {},
+    private val touchSlop: Float,
+    private val ignoredItems: List<Any> = emptyList(),
+) {
+    internal var draggingItemKey by mutableStateOf<Any?>(null)
+        private set
+
+    internal var interactingItemKey by mutableStateOf<Any?>(null)
+    internal var interactionSource = MutableInteractionSource()
+
+    private var draggingItemCumulatedOffset by mutableStateOf(Offset.Zero)
+    private var draggingItemInitialOffset by mutableStateOf(Offset.Zero)
+    internal var moved by mutableStateOf(false)
+    internal val draggingItemOffset: Offset
+        get() = draggingItemLayoutInfo?.let { item ->
+            draggingItemInitialOffset + draggingItemCumulatedOffset - item.offset.toOffset()
+        } ?: Offset.Zero
+
+    private val draggingItemLayoutInfo: LazyGridItemInfo?
+        get() = gridState.layoutInfo.visibleItemsInfo.firstOrNull { it.key == draggingItemKey }
+
+    internal var previousKeyOfDraggedItem by mutableStateOf<Any?>(null)
+        private set
+    internal var previousItemOffset = Animatable(Offset.Zero, Offset.VectorConverter)
+        private set
+
+    internal fun onDragStart(offset: Offset, shouldLongPress: Boolean) {
```

Comment:
```suggestion
    internal fun onDragStart(offset: Offset, shouldLongPressToDrag: Boolean) {
```

Reply:
In this case the parameter should tell us if we should long press the tab when the user holds the item past the `longPressTimeout` limit without any dragging gestures. This would then be countered by `onExitLongPress` if dragging is detected.

We could instead update the function naming, `onDragStart` doesn't sound like we're also capturing long press events. What do you think?

Reply:
Gotcha 👍 sure, did you have any names in mind?

Reply:
Not really, how about `onTouchSlopPassed` ?